[Backtracing] Add an indication that we're working on a backtrace.

al45tair opened the pull request:
When we crash, emit a message straight away that says we're working on a backtrace.  If starting the backtracer fails, report that also.

Finally, add a duration to the messages output by the backtracer, so that we can see how long it took.

rdar://118055527

al45tair committed:
[Backtracing] Add an indication that we're working on a backtrace.

When we crash, emit a message straight away that says we're working
on a backtrace.  If starting the backtracer fails, report that also.

Finally, add a duration to the messages output by the backtracer, so
that we can see how long it took.

rdar://118055527

al45tair commented:
@swift-ci Please smoke test

mikeash commented on stdlib/public/libexec/swift-backtrace/main.swift:
```diff
@@ -611,7 +652,13 @@ Generate a backtrace for the parent process.
+    if args.color {
+      write("\r\u{1b}[0K")
+    } else {
+      write("\r" + String(repeating:" ", count: 80) + "\r")
```
This hardcoded `80` is unfortunate, but it looks like this is just a fallback, so probably not a big deal?

al45tair replied:
Yeah, there isn't much we can do if we want to erase the message that was there before. We could choose instead to not erase it, and make the message different, maybe. Perhaps that would be OK for non-color mode?

al45tair replied:
I _suppose_ we could work out how long the message actually was, but if it's over the width of the terminal then you end up with odd output (since you get a line wrap).

mikeash replied:
That seems reasonable. I assume non-color mode is rare, outside of having your output piped somewhere. And if it's piped, trying to clear the message will just add garbage.

al45tair replied:
I fixed it so that we print a separate message in the non-color case, and also modified the code that deals with what happens when we don't find `swift-backtrace` so that _if you explicitly enable backtracing_, you'll get a warning that it's been disabled because we didn't have `swift-backtrace`.

mikeash reviewed (approved):
Nice!

al45tair committed:
[Backtracing] Massage the output a little in various cases.

We shouldn't try to erase the message when in non-color mode; mostly
that's intended for redirected output scenarios anyway, and in that case
we don't really want garbage in the output.

Also, improve the failure messages when the backtracer itself goes
wrong or can't be executed.

Finally, change the behaviour slightly so that *if* we're explicitly
enabled, *and* the backtracer path wasn't explicitly specified, *and*
we can't find the backtracer, we print a warning on start-up.  We
don't do that in any other case because we don't want spurious warnings
everywhere.

rdar://118055527

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
[Backtracing] Add some more tests.

Add tests for the new immediate on-crash message and the backtrace timing
message.

rdar://118055527

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
[Backtracing] Fix macOS test failure.

Apparently I forgot to code sign the test binaries.

rdar://118055527

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test
